### PR TITLE
Adds an ~5 minute guide to Numba.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
This is a WIP draft of a 5 minute guide to Numba. I'm not sure where it is going just yet, I've shoved it into the toctree for the documentation and written it in RST just for convenience so we can get a rendered version to look at. This will need adjusting if it does go into the formal documentation.

This guide is written in more colloquial style than the rest of the documentation (that is typically formal).  It is intended for new users/those who've just heard of Numba and want to see if it is a likely to be a good fit for them. It therefore focuses on how to get set up to use Numba, what Numba is good at and how to correctly assess whether it is doing a good job.

Feedback is very welcome.

TODO: 
- [ ] Fix up links based on whether this goes in docs or elsewhere.

File: docs/source/user/5minguide.rst [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+Numba is available as a `conda <https://conda.io/docs/>`_ package for the 
+`Anaconda Python distribution <https://www.anaconda.com/>`_::
+
+  $ conda install numba
+
+Numba also has wheels available::
+
+  $ pip install numba
+
+Numba can also be compiled from source (`instructions <LINK>`_)
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: ", although we do not recommend it for first-time Numba users."

File: docs/source/user/5minguide.rst [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+Numba is often used as a core package so its dependencies are kept to an
+absolute minimum, however, extra packages can be installed as follows to provide
+additional functionality:
+
+* ``scipy`` - enables support for compiling ``numpy.linalg`` functions
+* ``colorama`` - enables support for color highlighting in backtraces/error
+  messages
+* ``pyyaml`` - enables configuration of Numba via a YAML config file.
+* ``icc_rt`` - allows the use of the Intel SVML (high performance short vector
+  math library, x86_64 only).
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: Link to performance guide on icc_rt, so people see how to install it.

File: docs/source/user/5minguide.rst [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+functions to demonstrate what works well and what does not.
+
+Numba works well on code that "looks like" this::
+
+    from numba import jit
+    import numpy as np
+
+    x = np.arange(100).reshape(10, 10)
+
+    @jit
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: Although you discuss nopython mode below, I would *start* the example with nopython mode and then rephrase the section below to explain why you do this.  I'd rather show best general practices, and then explain when to break them, than vice versa.

File: docs/source/user/5minguide.rst [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    def loop_lift(x):
+        a = object() # Numba runs this in the interpreter
+        acc = 0
+        for i in x:  # Numba will compile this loop to machine code!
+            acc +=i
+        return a, acc
+
+    print(loop_lift(x))
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: At this point, I consider loop lifting to be an advanced Numba feature that we can leave out of a quickstart guide.

File: docs/source/user/5minguide.rst [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        df += 1                        # Numba doesn't understand what this is
+        return df.cov()                # or this!
+
+    print(use_pandas(x))
+
+Note that Pandas is not understood by Numba and as a result Numba would simply
+run this code via the interpreter but with the added cost of the Numba internal
+overheads!
+
+The ``nopython`` mode *gotcha*.
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: Retitle this as "What is ``nopython`` mode?"

File: docs/source/user/5minguide.rst [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+-------------------------------
+Pretty much all first time users are caught by this so it deserves a mention...
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: I would shorten this section and write it assuming nopython mode is the standard way to use `@jit`, mention briefly that Numba has limited support for mixed mode via loop lifting and maybe link to somewhere in the docs that give more information.

File: docs/source/user/5minguide.rst [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+
+Numba works well on code that "looks like" this::
+
+    from numba import jit
+    import numpy as np
+
+    x = np.arange(100).reshape(10, 10)
+
+    @jit
+    def go_fast(a): # Function is compiled and runs in machine code
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: Change this comment to `# Function is compiled to machine code when called the first time`

File: docs/source/user/5minguide.rst [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+This, for example prints::
+
+    Elapsed (with compilation) = 0.33030009269714355
+    Elapsed (after compilation) = 6.67572021484375e-06
+
+A good way to measure the impact Numba JIT has on your code is to time execution
+using the `timeit <https://docs.python.org/3/library/timeit.html>`_ module
+functions, these measure multiple iterations of execution and, as a result,
+can be made to accommodate for the compilation time in the first execution.
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: Show example of doing `%timeit` in the Jupyter Notebook / IPython as well.

File: docs/source/user/5minguide.rst [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+---------------
+Assuming Numba can operate in ``nopython`` mode, or at least compile some loops,
+it will target compilation to your specific CPU. Speed up varies depending on
+application but can be one to two orders of magnitude. Numba has a
+`performance guide <LINK>`_ that covers common options for gaining extra
+performance.
+
+How does Numba work?
+--------------------
+Glossing over a huge amount of detail, Numba reads the Python bytecode for a
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: Don't need "Glossing over a huge amount of detail".  😄

File: docs/source/user/5minguide.rst [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+decorated function and combines this with information about the types of the
+input arguments to the function. It then forms a Numba specific internal
+representation(IR) of the function and translates this to LLVM IR, it then
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: We can go even simpler on this discussion.  "It analyzes and optimizes your code, and finally uses the LLVM compiler library to generate a machine code version of your function, tailored to your CPU capabilities."

File: docs/source/user/5minguide.rst [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+Extra options available in some decorators:
+
+* ``parallel = True`` - enable `automatic parallelization <LINK NB PARFORS>`_ of
+  the function.
+* ``fastmath = True`` - enable `fast-math <LINK NB fast-math>`_ behaviour for
+  the function.
+
+ctypes/cffi/cython interoperability:
+
+* ``cffi`` - `CFFI  <LINK NB CFFI DOCS>`_ functions are supported.
```
Comment (seibert) [inline docs/source/user/5minguide.rst]: I would phrase these as "Calling X functions is supported in nopython mode."

**Comment by seibert** [comment]
I think we want to pair this with a notebook of code examples, and then embed a link near the top of the guide to https://mybinder.org/ so they can try them out in their browser.

**Comment by stuartarchibald** [comment]
Thanks for the review @seibert. I'm happy for this to be superseded by the outcome of the front-page description discussion on gitter, or for some amalgamation to happen if that's convenient, or this can stand as is? Thoughts?

**Comment by stuartarchibald** [comment]
The only remaining broken link is to ROC which is in PR https://github.com/numba/numba/pull/3178. Should the example code/notebooks etc be done in a separate PR to this one?

**Comment by seibert** [comment]
Examples and code can be a separate PR.

**Comment by stuartarchibald** [comment]
ok, I think that's a good separation of work. I think I've addressed all the comments, so pending merge of ROC docs so I can update that link (and any other further review comments) I think this is sufficiently complete to consider merge :)

**Comment by seibert** [comment]
ROC docs are merged now.

**Comment by seibert** [comment]
Ready to merge as soon as one of the Travis CI build_doc configurations pass...

**Comment by seibert** [comment]
Doc build passed.  Merging

**Pull request merged**